feat: adds motd to the ui

flyinghermit opened the pull request:
When MOTD is set, the message is shown before login screen (`tsh` does the same). User's are required to click confirmation button before transitioning to login screen. 

MOTD in teleport is configurable with `auth_service.message_of_the_day ` in `teleport.yaml` config. See docs for setup instruction: https://goteleport.com/docs/server-access/guides/ssh-pam/#display-a-message-of-the-day-motd-with-teleport

Closes https://github.com/gravitational/teleport/issues/13724

flyinghermit committed:
feat: adds motd to the ui

kimlisa commented on web/packages/teleport/src/Login/Login.tsx:
```diff
@@ -41,25 +42,35 @@ export function Login({
-        primaryAuthType={primaryAuthType}
-        privateKeyPolicyEnabled={privateKeyPolicyEnabled}
-      />
+      {motd && showMOTD ? (
```
add behavioral test?

sshahcodes replied:
Added tests in https://github.com/gravitational/teleport/pull/27813/commits/36314beaa6c0cc0ff2c79b90bf2f0748697de19c

kimlisa reviewed (commented):
can you manually test that this works for both the open source and enterprise version?

and should we be preserving the users first `acknowledge`? so it doesn't keep popping up every time they log in?

kimlisa commented on lib/web/apiserver.go:
```diff
@@ -1397,6 +1397,7 @@ func (h *Handler) getWebConfig(w http.ResponseWriter, r *http.Request, p httprou
 			PreferredLocalMFA:  cap.GetPreferredLocalMFA(),
 			LocalConnectorName: localConnectorName,
 			PrivateKeyPolicy:   cap.GetPrivateKeyPolicy(),
+			MOTD:               cap.GetMessageOfTheDay(),
```
there's a test for this that we should probably add to

sshahcodes replied:
added in https://github.com/gravitational/teleport/pull/27813/commits/af6c5b6171a91e93670cf67a0b2ac365f12c97c7

zmb3 commented on web/packages/teleport/src/Login/MOTD/MOTD.tsx:
```diff
@@ -0,0 +1,41 @@
+    <Card bg="levels.surface" my={6} mx="auto" width="464px">
+      <Box p={6}>
+        <Text typography="h2" mb={3} textAlign="center" color="text.main">
+          Message of the day!
```
Do we need a title like this? If so, should it end with an `!`?

sshahcodes replied:
My thinking was to title the component and be verbose on what we are displaying below as motd message. I think that `!` is not necessary though. 

@roraback what do you think about keeping title?

roraback replied:
I would lean toward removing it. I do like having _a_ title, but because it isn't customizable and different customers may use this feature in quite different ways, I think it's probably best to just leave it out. In particular, I'm imagining a situation in which it's the same message for many days, and the "Message of the Day" title becomes quite ironic for users who see it day after day.

sshahcodes replied:
Noted. Thanks for the input. Removed title in https://github.com/gravitational/teleport/pull/27813/commits/0995d036999bc34f211b5e399c36350ad95d39ef

zmb3 commented on web/packages/teleport/src/config.ts:
```diff
@@ -280,6 +282,10 @@ const cfg = {
     return cfg.auth ? cfg.auth.preferredLocalMfa : null;
   },
 
+  getMOTD() {
+    return cfg.auth ? cfg.auth.motd : null;
+  },
```
+1

flyinghermit committed:
Merge branch 'master' into sshah/motd-ui

sshahcodes commented on web/packages/teleport/src/Login/Login.tsx:
```diff
@@ -41,25 +42,35 @@ export function Login({
   privateKeyPolicyEnabled,
+  motd,
 }: State) {
+  const [showMOTD, setShowMOTD] = useState<boolean>(true);
```
Completely missed enterprise `Login.tsx`. Thanks for pointing out.

flyinghermit committed:
address review suggestions:
- update MOTD to Motd
- moved motd state to useLogin
- added behaviour tests

flyinghermit committed:
add motd test to apiserver_test.go webconfig. update snapshot test to address motd warning

sshahcodes commented:
> can you manually test that this works for both the open source and enterprise version?

Tested in both oss and enterprise. enterprise PR here https://github.com/gravitational/teleport.e/pull/1654

> and should we be preserving the users first acknowledge? so it doesn't keep popping up every time they log in?

I don't think we would want to preserve acknowledge state (even though UX wise, users may find it annoying).  MOTDs are meant to be shown before each login. Admins may also update MOTD frequently.

flyinghermit committed:
git mv to update MOTD to Motd

kimlisa commented on web/packages/teleport/src/Login/Login.test.tsx:
```diff
@@ -113,3 +113,35 @@ test('login with private key policy enabled through role setting', async () => {
+test('show motd only if motd is set', async () => {
+  // default login form
+  render(<Login />);
+  expect(screen.getByPlaceholderText(/username/i)).toBeInTheDocument();
```
i'd also check that there was no 'message of the day' in the document

kimlisa commented on web/packages/teleport/src/Login/useLogin.ts:
```diff
@@ -74,6 +75,12 @@ export default function useLogin() {
     history.push(ssoUri, true);
   }
 
+  const [showMotd, setShowMotd] = useState(!!motd);
```
we should group react states with other react states (near the top)

kimlisa commented on web/packages/teleport/src/Login/useLogin.ts:
```diff
@@ -74,6 +75,12 @@ export default function useLogin() {
 
+  const [showMotd, setShowMotd] = useState(!!motd);
+
+  const acknowledgeMotd = () => {
+    setShowMotd(false);
+  };
```
nit: i'd prefer using `function acknowledgeMotd() {}` over arrow func (if you don't need the features of an arrow func), simply b/c it's easier to read imo (and other func in this comp is defined that way)

flyinghermit committed:
multiple fix:
- add unmount test
- remove motd title
- group states together in useLogin
- update arrow func to classic js func

flyinghermit committed:
Merge branch 'master' into sshah/motd-ui

flyinghermit committed:
remove unused waitForElementToBeRemoved

flyinghermit committed:
Merge branch 'sshah/motd-ui' of github.com:gravitational/teleport into sshah/motd-ui
